jax/scipy/optimize/bfgs_minimize.py

```diff
@@ -0,0 +1,127 @@
+from jax.lax import while_loop
+from .line_search import line_search
+import collections
+
+BFGSResults = collections.namedtuple(
+    'BFGSResults', [
+        'converged',  # bool, True if minimization converges
+        'failed',  # bool, True if line search fails
+        'k',  # The number of iterations of the BFGS update.
+        'nfev',  # The total number of objective evaluations performed.
+        'ngev',  # total number of jacobian evaluations
+        'nhev',  # total number of hessian evaluations
+        'x_k',  # A tensor containing the last argument value found during the search. If the search converged, then
+        # this value is the argmin of the objective function.
+        'f_k',  # A tensor containing the value of the objective
+        # function at the `position`. If the search
+        # converged, then this is the (local) minimum of
+        # the objective function.
+        'g_k',  # A tensor containing the gradient of the
+        # objective function at the
+        # `final_position`. If the search converged
+        # the max-norm of this tensor should be
+        # below the tolerance.
+        'H_k'  # A tensor containing the inverse of the estimated Hessian.
+    ])
+
+
+def bfgs_minimize(func, x0, options=None):
+    """
+        Minimisation with BFGS. Inexact line search is performed satisfying both Wolfe conditions
+        (strong version of second one).
+        Args:
+            func: function to minimise, should take a single flat parameter, similar to numpy
+            x0: initial guess of parameter
+            options: A dictionary of solver options. All methods accept the following generic options:
+                maxiter : int or None which means inf
+                    Maximum number of iterations to perform. Depending on the
+                    method each iteration may use several function evaluations.
+                analytic_initial_hessian: bool,
+                    whether to use JAX to get hessian for the initialisation, otherwise it's eye.
+                g_tol: float
+                    stopping criteria is |grad(func)|_2 \lt g_tol
+                ls_maxiter: int or None which means inf
+                    Maximum number of line search iterations to perform
+
+        Returns: BFGSResults
+
+    """
+    if options is None:
+        options = dict()
+    maxiter: int = options.get('maxiter', None)
+    analytic_initial_hessian: bool = options.get('analytic_initial_hessian', True)
+    g_tol: float = options.get('g_tol', 1e-8)
+    ls_maxiter: int = options.get('ls_maxiter', 50)
+
+    state = BFGSResults(converged=False,
+                        failed=False,
+                        k=0,
+                        nfev=0,
+                        ngev=0,
+                        nhev=0,
+                        x_k=x0,
+                        f_k=None,
+                        g_k=None,
+                        H_k=None)
+
+    if maxiter is None:
+        maxiter = np.inf
```

Comment:
This may lead to pytype errors because `maxiter` is declared as an int above.

Reply:
Typing, being only a weak suggesting in python, shouldn't have an effect on this evaluation if I'm not mistaken? The `is` should only check if objects are the same, whereas == would also compare value equality. Let me know if I'm misunderstanding.